src/core/meshcop/dataset.cpp

```diff
@@ -590,10 +591,11 @@ otError Dataset::ApplyConfiguration(Instance &aInstance) const
             memcpy(networkName.m8, name->GetNetworkName(), name->GetLength());
             networkName.m8[name->GetLength()] = '\0';
 
-            if (strcmp(networkName.m8, mac.GetNetworkName()) != 0)
+            if (!notifier.HasSignaled(OT_CHANGED_THREAD_NETWORK_NAME) ||
```

Comment:
It's somewhat odd that some configuration patterns require the `HasSignaled()` here and others perform that internally as part of the `Set()` method.  Can we make this more consistent?

Reply:
Changed the model in the new puhsed commit. All the `OT_CHANGED_<>` signals are emitted from corresponding `Set` methods. 

@pvanhorn, with this change we can get `OT_CHANGED_THREAD_CHANNEL` when user sets the channel (possibly before device is fully attached and creates a valid Active dataset). When we get `OT_CHANGED_THREAD_CHANNEL`, we can get the new channel from `otLinkGetChannel()` API.